### DISCUSSION TITLE
How to pass javascript function to eCharts

### DISCUSSION THREAD

**Discussion by italanchan** [zauberzeug/nicegui#1588] [Q&A]
### Question

Hi, 

Try to convert the echart example to Nicegui:

https://echarts.apache.org/examples/en/editor.html?c=pie-half-donut

The code:
```py
from nicegui import ui

option = {
  "tooltip": {
    "trigger": 'item'
  },
  "legend": {
    "top": '5%',
    "left": 'center',
    # // doesn't perfectly work with our tricks, disable it
    "selectedMode": False
  },
  "series": [
    {
      "name": 'Access From',
      "type": 'pie',
      "radius": ['40%', '70%'],
      "center": ['50%', '70%'],
      # // adjust the start angle
      "startAngle": 180,
      "label": {
        "show": True,
        "formatter": "formatter(param) { \
          return param.name + ' (' + param.percent * 2 + '%)'; \
        }"
      },
      "data": [
        { "value": 1048, "name": 'Search Engine' },
        { "value": 735, "name": 'Direct' },
        { "value": 580, "name": 'Email' },
        { "value": 484, "name": 'Union Ads' },
        { "value": 300, "name": 'Video Ads' },
        {
        #   // make an record to fill the bottom 50%
          "value": 1048 + 735 + 580 + 484 + 300,
          "itemStyle": {
            # // stop the chart from rendering this piece
            "color": 'none',
            "decal": {
              "symbol": 'none'
            }
          },
          "label": {
            "show": False
          }
        }
      ]
    }
  ]
}

echart = ui.echart(option)

ui.run()
```

The result is:

![image](https://github.com/zauberzeug/nicegui/assets/10571926/1428c5cf-bfc5-47d5-852f-83a9537680a9)

The label cannot be parse.

How to pass javascript func to option in Nicegui.

**Comment by natankeddem** [comment]
There is an active PR to get this functionality working; https://github.com/zauberzeug/nicegui/pull/1579.

**Reply by ed2050** [reply]
Very nice, I got this to work.  Looks like it was added [in a recent release](https://nicegui.io/documentation/e_chart).  Here's what I did if it helps others (had trouble when I started this post, but then figured it out):

# Setup
I keep my js in a separate file helpers.js.  My nicegui app loads helpers.js with ui.add_head_html.  Why do it this way?
- I have some js functions that don't fit easily in one line 
- more convenient to have js in separate file with proper indentation, syntax highlighting, etc

Then I just have to call my js functions from the :formatter property.  

# Examples

I use a js func fmt_size_js to customize how size values are displayed.  It shows filesize as "3 MB" or "12 GB" instead of plain bytes.  This is currently used with both ui.table and ui.echart.

## Table
```py
    columns = [
        {
            'name'  : 'size' , 'label' : 'Size' , 'field' : 'size' ,
            ':format' : 'fmt_size_js' ,
        } ,
ui.table (columns = columns, rows = [])
# works, shows formatted size in KB, MB, or GB 
```

## Echart
```py
piechart = {
    'tooltip' : {
        'trigger'   : 'item' ,
        #'formatter' : '{b} : {d}% <br /> {c}' ,  # works fine, but limited
        ':formatter' : 'fmt_pie_label' , 
    },
ui.echart (piechart)
# fmt_pie_label takes the value arg from params and converts it with fmt_size_js
```

Very nicely done!  Big thanks to @thetableman @natankeddem @rodja @falkoschindler and anyone else who worked on this!

# Issues
The biggest issue was that my helper funcs weren't being picked up.  Browser console would complain that fmt_size wasn't defined.  ui.run () doesn't seem to pickup changes to js files, even though I added *.js to uvicorn_reload_includes.  Not sure why, will look into it later.  Restarting my app fixed it.

**Reply by vickorian** [reply]
@ed2050 know this is a bit of an old post, but how are you importing from file using the add_html_header?

**Reply by falkoschindler** [reply]
@vickorian Sometimes we do it like this:
https://github.com/zauberzeug/nicegui/blob/eac5a5faa9fbd8823a6b76784a76fce618fd7afc/examples/webserial/main.py#L30

**Comment by ed2050** [comment]
> @ed2050 know this is a bit of an old post, but how are you importing from file using the add_html_header?

like this: 
```py
# first, tell server how to serve my js files
app.add_static_files ('/js' , '/some/local/path/')

# in page method, add js file to <head> of created page  
ui.add_head_html (f'<script type="text/javascript" src="/js/helpers.js"> </script>')

# now the page will load http://myserver/js/helpers.js from /some/local/path/helpers.js
```

add_head_html is tricky.  It only works when called before any of the page is rendered or client connection is established (eg `await client.connected ()`).  Otherwise you just get a silent failure iirc.

It's easier to use `add_body_html` like falko says.  For most js and css, add_body works fine.   In my case, I think there was some js I wanted to run before the page finished loading, so add_head_html was needed.  Also, the docs probably suggested add_head_html at the time, though that may have changed by now.

**Reply by falkoschindler** [reply]
@ed2050 I think nowadays `add_head_html` can be called even after the client has connected. The new HTML is simply added at the end of `<head>`:

https://github.com/zauberzeug/nicegui/blob/8828a56ff4c982480932c5c36558feda09207cb8/nicegui/functions/html.py#L17-L18

**Reply by ed2050** [reply]
Cool.  You might want to update the [code comment](https://github.com/zauberzeug/nicegui/blob/8828a56ff4c982480932c5c36558feda09207cb8/nicegui/functions/html.py#L8) then.  It still says:

- Note that this function can only be called before the page is sent to the client.
